src/tiled/abstractobjecttool.cpp

```diff
@@ -209,6 +209,42 @@ void AbstractObjectTool::saveSelectedObject()
         mapDocument()->saveSelectedObject(name, groupIndex);
 }
 
+void AbstractObjectTool::detachSelectedObjects()
+{
+    MapDocument *currentMapDocument = mapDocument();
+    QList<MapObject *> templateInstances;
+
+    /**
+     * Stores the unique tilesets used by the templates
+     * to avoid creating multiple undo commands for the same tileset
+     */
+    QSet<SharedTileset> sharedTilesets;
+
+    for (MapObject *object : mapDocument()->selectedObjects()) {
+        if (object->templateObject()) {
+            templateInstances.append(object);
+
+            if (Tile *tile = object->cell().tile())
+                sharedTilesets.insert(tile->tileset()->sharedPointer());
+        }
+    }
+
+    auto changeMapObjectCommand = new  QUndoCommand(
+                QCoreApplication::translate("Undo Commands",
+                                            "Detach %n Template Instance(s)",
+                                            nullptr, templateInstances.size()));
+
+    // Add any missing tileset used by the templates to the map map before detaching
+    for (SharedTileset sharedTileset : sharedTilesets) {
+        if (!currentMapDocument->map()->tilesets().contains(sharedTileset))
+            new AddTileset(currentMapDocument, sharedTileset, changeMapObjectCommand);
+    }
+
+    new DetachObject(currentMapDocument, templateInstances, changeMapObjectCommand);
```

Comment:
Please rename this class to `DetachObjects` now that it handles multiple.

Note that you could add support for child undo commands to `DetachObjects` by calling the parent undo/redo methods as well, in which case you can make the `AddTileset` commands children and don't need the separate `changeMapObjectCommand`.

Reply:
I'm not sure what do you mean so this might not be related, I created the `AddTileset` undo action before the `DetachObjects` action to make sure that any tile replacement works as expected, otherwise the `GidMapper` won't be able to to map the tile to the correct gid. The initial QUndo command contains the `AddTileset` Actions and eventually the `DetachObjects` action as well.

Reply:
The order of execution of the undo commands doesn't matter in this case, because they do not depend on each other for their state modifications and they can't be undone or redone separately (they form a single entry in the undo history).

See other created instances of `AddTileset`, like in `AbstractObjectTool::changeTile`, for what I meant with attaching it as a child to the main undo command (and `ChangeMapObjectsTile::undo/redo`). I just meant that doing it that way avoids the need to create an additional `QUndoCommand` instance.

Reply:
Hmm, I initially it was changeTile, specifically like this:
```
    auto changeMapObjectCommand = new DetachObjects(currentMapDocument,
                                                    templateInstances);

    for (SharedTileset sharedTileset : sharedTilesets) {
        if (!currentMapDocument->map()->tilesets().contains(sharedTileset))
            new AddTileset(currentMapDocument, sharedTileset, changeMapObjectCommand);
    }

    currentMapDocument->undoStack()->push(changeMapObjectCommand);
```

But the gid of the detached instance became 0, I suspected that the order mattered so added the empty udnoCommand.

Reply:
It's because it won't magically work, you'd need to change `DetachObjects` to call its superclass in the undo/redo functions, like `ChangeMapObjectsTile::undo/redo`.

Reply:
Alright, I've updated it. I thought I had to manually order them on creation.